packages/core/deprecations/core-deprecations-server-internal/src/routes/post_validation_handler.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import type { InternalCoreUsageDataSetup } from '@kbn/core-usage-data-server-internal';
+import type { CoreKibanaRequest } from '@kbn/core-http-router-server-internal';
+import type { InternalHttpServiceSetup } from '@kbn/core-http-server-internal';
+import { isObject } from 'lodash';
+import { RouteDeprecationInfo } from '@kbn/core-http-server/src/router/route';
+import { buildApiDeprecationId } from '../deprecations';
+
+interface Dependencies {
+  coreUsageData: InternalCoreUsageDataSetup;
+  http: InternalHttpServiceSetup;
+}
+
+/**
+ * listens to http post validation events to increment deprecated api calls
+ * This will keep track of any called deprecated API.
+ */
+export const registerApiDeprecationsPostValidationHandler = ({
+  coreUsageData,
+  http,
+}: Dependencies) => {
+  http.registerOnPostValidation(createRouteDeprecationsHandler({ coreUsageData }));
+};
+
+export function createRouteDeprecationsHandler({
+  coreUsageData,
+}: {
+  coreUsageData: InternalCoreUsageDataSetup;
+}) {
+  return (req: CoreKibanaRequest, { deprecated }: { deprecated?: RouteDeprecationInfo }) => {
+    if (deprecated && isObject(deprecated)) {
+      const counterName = buildApiDeprecationId({
+        routeMethod: req.route.method,
+        routePath: req.route.path,
```

Comment:
I tested locally and with deprecated routes with params path will not be what we want. E.g.:

```
POST /api/alerts/alert/{id?}
```

Will actually be

```
POST /api/alerts/alert/test
```

So the counter won't go to the registered route. I got around this by exposing the actual router route on the request object like this:

```diff
diff --git a/packages/core/http/core-http-router-server-internal/src/request.ts b/packages/core/http/core-http-router-server-internal/src/request.ts
index 6d4da18d0a3..b56b4764427 100644
--- a/packages/core/http/core-http-router-server-internal/src/request.ts
+++ b/packages/core/http/core-http-router-server-internal/src/request.ts
@@ -256,7 +256,12 @@ export class CoreKibanaRequest<
     } = request.route?.settings?.payload || {};
 
     // the socket is undefined when using @hapi/shot, or when a "fake request" is used
-    const socketTimeout = isRealRawRequest(request) ? request.raw.req.socket?.timeout : undefined;
+    let socketTimeout: undefined | number;
+    let routePath: undefined | string;
+    if (isRealRawRequest(request)) {
+      socketTimeout = request.raw.req.socket?.timeout;
+      routePath = request.route.path;
+    }
     const options = {
       authRequired: this.getAuthRequired(request),
       // TypeScript note: Casting to `RouterOptions` to fix the following error:
@@ -291,6 +296,7 @@ export class CoreKibanaRequest<
 
     return {
       path: request.path ?? '/',
+      routePath,
       method,
       options,
     };

```

And then using here like:

```diff
diff --git a/packages/core/deprecations/core-deprecations-server-internal/src/routes/post_validation_handler.ts b/packages/core/deprecations/core-deprecations-server-internal/src/routes/post_validation_handler.ts
index 52e261e3da9..914be43600a 100644
--- a/packages/core/deprecations/core-deprecations-server-internal/src/routes/post_validation_handler.ts
+++ b/packages/core/deprecations/core-deprecations-server-internal/src/routes/post_validation_handler.ts
@@ -36,10 +36,10 @@ export function createRouteDeprecationsHandler({
   coreUsageData: InternalCoreUsageDataSetup;
 }) {
   return (req: CoreKibanaRequest, { deprecated }: { deprecated?: RouteDeprecationInfo }) => {
-    if (deprecated && isObject(deprecated)) {
+    if (deprecated && isObject(deprecated) && req.route.routePath) {
       const counterName = buildApiDeprecationId({
         routeMethod: req.route.method,
-        routePath: req.route.path,
+        routePath: req.route.routePath,
         routeVersion: req.apiVersion,
       });

```

up to you how you want to approach it!